langchain: add RankLLM Reranker

16BitNarwhal opened the pull request:
Integrate RankLLM reranker (https://github.com/castorini/rank_llm) into LangChain

An example notebook is given in `docs/docs/integrations/retrievers/rankllm-reranker.ipynb`

16BitNarwhal committed:
init rankllm integration

16BitNarwhal committed:
Merge branch 'langchain-ai:master' into integrate-rankllm

16BitNarwhal committed:
format

16BitNarwhal committed:
Merge branch 'langchain-ai:master' into integrate-rankllm

16BitNarwhal committed:
Merge branch 'langchain-ai:master' into integrate-rankllm

16BitNarwhal committed:
integration test + small fix

16BitNarwhal committed:
Merge branch 'integrate-rankllm' of github.com:16BitNarwhal/langchain into integrate-rankllm

16BitNarwhal committed:
add RankZephyr example

16BitNarwhal committed:
Merge branch 'master' into integrate-rankllm

ronakice commented:
Great work @16BitNarwhal on integrating this to Langchain. Hopefully this brings in more interest in high precision and effective listwise paradigm for reranking models like RankZephyr!

hwchase17 commented on libs/langchain/langchain/retrievers/document_compressors/rankllm_rerank.py:
```diff
@@ -0,0 +1,127 @@
+
+                    from langchain.llms.openai import OpenAI
+
+                    llm = OpenAI(
```
you are importing just to access the api key? lets not do that, surely there has to be a more direct way

hwchase17 commented on libs/langchain/langchain/retrievers/document_compressors/rankllm_rerank.py:
```diff
@@ -0,0 +1,127 @@
+from langchain_core.documents import Document
+from langchain_core.pydantic_v1 import Extra, Field, PrivateAttr, root_validator
+
+from langchain.callbacks.manager import Callbacks
```
should probably import from core?

hwchase17 reviewed (commented):
this should go in langchain-community, not langchain, as it is a third party integration

16BitNarwhal committed:
move rankllm to community

16BitNarwhal committed:
fix get openai apikey

16BitNarwhal committed:
Merge branch 'integrate-rankllm' of github.com:16BitNarwhal/langchain into integrate-rankllm

16BitNarwhal committed:
Merge branch 'master' into integrate-rankllm

16BitNarwhal committed:
move test and format

16BitNarwhal commented:
Hey @hwchase17, thanks for the feedback! Made the following changes:
- Moved rankllm to community along with the integration test and updated example notebook
- Imported from langchain_core instead of langchain for callbacks
- Directly get openai API key from env instead of importing openai

Also, I'm a bit unsure about something. I noticed that [flashrank](https://github.com/langchain-ai/langchain/blob/893f06b5de4dccfc0c388a5e0f99715be27789dd/libs/langchain/langchain/retrievers/document_compressors/flashrank_rerank.py#L24) doesn't seem to be located in community. I see they support models like RankZephyr, which was built and is maintained by RankLLM team. Would it also be more appropriate for FlashRank to be in community?

hwchase17 commented:
> Also, I'm a bit unsure about something. I noticed that [flashrank](https://github.com/langchain-ai/langchain/blob/893f06b5de4dccfc0c388a5e0f99715be27789dd/libs/langchain/langchain/retrievers/document_compressors/flashrank_rerank.py#L24) doesn't seem to be located in community. I see they support models like RankZephyr, which was built and is maintained by RankLLM team. Would it also be more appropriate for FlashRank to be in community?

Yes it should be, thanks for flagging. we will move

16BitNarwhal commented:
Sounds good. And please let me know if there's anything else I need to do to get this merged 😁

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 commented:
> Sounds good. And please let me know if there's anything else I need to do to get this merged 😁

some test/lint issues, i can try to resolve

16BitNarwhal committed:
fixes

16BitNarwhal committed:
Merge branch 'master' into integrate-rankllm

hwchase17 committed:
Merge branch 'master' into integrate-rankllm

hwchase17 committed:
Merge branch 'master' into integrate-rankllm

hwchase17 committed:
Merge branch 'master' into integrate-rankllm

hwchase17 committed:
cr
